mantidproject/mantid
# Preserve height when FWHM of BackToBackExponential changed

**Description of work.**
Previously the area (first parameter) was kept constant when changing height such the peak height would increase and prevent user from accessing the height slider.
Now the height is reset when the FWHM is changed. The height was previously calculated by evaluating the function at X0, but ab initio the user will not be able to find X0 accurately as it depends on A,B,S etc. so this has been replaced with an analytic equation for the maximum of the function (which is typically what the user will interpret as the height).
In addition equations to map FWHM to S  and vice-versa have been added. Now the estimated FWHM provided  a more realistic guess for S based on the other parameters A and B and the changes to FWHM make more appropriate changes to S. The equations obey the limiting behaviour of the function (small S the FWHM is that of the BackToBackExponential - w0 in the code - and large S is a Gaussian) and are good to <3% for typical parameters. It is now no longer possible to set S<1e-6 with the sliders (arbitrary non-zero small number that essentially returns the unconvoluted BackToBackExponential - the function is not well behaved for S=0).

Note the analytical equations for the height and FWHM were determined empirically from simulated datasets for parameters in the range
- 0.005 < A < 0.12 muS^-1
- 0.005 < B < 0.12 muS^-1
- 0.1 < S < 250 muS

**To test:**
(1) Load spectra or focussed data to test (note that this function always requires a good guess for A and B so pick an instrument for which these values are calibrated in the parameter.xml - or use this data from ENGIN-X [ENGINX_193749_bank_1.zip](https://github.com/mantidproject/mantid/files/5459011/ENGINX_193749_bank_1.zip) 
(2) Open the fit browser and right-click on the plot > Select Peak Type > BackToBackExponential
(3) Add the peak
(4) On the fit browser click Display > Plot Guess
(5) Move the sliders for FWHM and height - the height should stay roughly constant for realistic FWHM. 
- At large S it decreases slightly (the equation is only good to a few percent and the error accumulates - but it still is much easier to find better starting values)
- As the FWHM is decreased below the intrinsic back to back exponential width (S tends to 0) then the function should saturate at S=1e-6 (see above)

(6) Perform a fit just to test the function still behaves (though the changes shouldn't effect the evaluation of the function in a fit)

Fixes #29831 





---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
